### DISCUSSION TITLE
PIO micropython pico simple jmp() stops PIO state machine???

### DISCUSSION THREAD

**Discussion by budley** [micropython/micropython#12770] [RP2040 / Pico]
I was trying to write a simple PIO routine to detect a break sensor on a GPIO pin; calling back to micropython using an IRQ, but it just never seemed to trigger.
I boiled the code down to bare basics to try to figure out what was going on.
def irq_test():
    irq(1)
    jmp("continueloop")
    label("continueloop")
This calls the interrupt hooked up function once and then never again. Maybe the state machine stops???
If I remove the jmp, ie.
def irq_test():
    irq(1)
    label("continueloop")
It calls the interrupt hooked up function repeatedly, as you would expect. (the reason for the jmp is that that is the place where the jmp test on GPIO input pin would be, but I was trying to boil things down to see why it wasn't working) 
I have stuffed nop() [31]s in everywhere and that made no difference.

Does a jmp() do more than simply move execution to another point in the code?
Is it a bug in micropython implementation?
Am I missing something fundamental? (for sure yes, but specifically in my understanding here :-) ?!)
Thanks

**Comment by budley** [comment]
I realised that I had failed to put nop() in one place:
def irq_test():
    irq(1)
    jmp("continueloop")
    label("continueloop")
    nop()

This now loops repeatedly. This behaviour strikes me as potentially a micropython bug. Maybe there's a technical explanation for why it does this, but to a newbie this behaviour looks just plain wrong :-(

Any thoughts anybody??? Thanks

**Comment by mendenm** [comment]
You may want to look very hard at the discussion of PIO instructions in the rp2040 manual. I haven't refreshed my memory right now, but IIRC there is a weird interaction between jmp targets and the wrap target at  the end of a program.  I don't think you are seeing a micropython bug, but a PIO subtlety.

**Comment by redhead-p** [comment]
Yes - have a look at .wrap in rp2040 manual p317 in the online version.  This marks the end of the PIO code and it will automatically wrap the beginning when the last instruction, so marked, is executed - but not if the last instruction is an unconditional jump or conditional jump where the condition is satisfied. In that case the jump target takes precedence. '.wrap' or wrap() in MicroPython is automatically applied to the last instruction if you don't specify it.  Effectively the wrap marker is stored outside the program so that you don't need to use a valuable memory location to store the jump back to the beginning.  'nop()' isn't a jump so when wrap() is applied to it automatically it behaves as expected.

**Reply by budley** [reply]
Absolutely fascinating, thanks for taking the time to help me out and with such a specific reference. Knowing that has helped me solve my conundrum and now I know that, it is in fact a useful trick to be able to use that fact to wrap or not wrap. My (toy) application is very non-demanding so I'm not fussed by wasted cycles, in fact there is totally no need to use PIO... but you know how it goes, the resource is there so it begs to be utilised (I am giving a very old Boeing 767 flight deck chronometer new life by taking the non-working guts out and replacing the displays with 2x LCDs and using a stepper motor to drive the chronometer hand driven by a Pico. The stepper is the part I am swapping to be driven by PIO, 513 steps, driving it in short bursts to give "second steps"... you have helped with the bit on how to "home it" (using a breakthrough sensor) totally under PIO control. With pure micropython control over that I have to pause the digital clock whilst doing that or else the homing stutters as it redraws the display on interrupt (I tried using the second core for homing but that didn't quite stop that, I guess because there is some overhead in the timer interrupt even if on a different core) ).  Maybe my next step will be to consider if using the PIO controller would allow me to use a 'normal motor' to drive the chrono hand, since in the original instrument it is completely smooth but I didn't think possible to drive at an exact 1 rev per minute speed... but with your PIO help I'm now thinking that might indeed be possible (my first Pico/microcontroller project so I'm learning every step of the way). Sorry for rambling, thought the context might be of interest, thanks again.
